Feature_MultiPointers

Railboy opened the pull request:
Overview
---
Added optional EventOrigin and SourceType members to BaseInputEventData.

Updated InputManager to initialize (most) input events with these arguments.

InputManager determines input source type on input source detected via DetermineInputSourceType.

InputManager can add / remove EventOrigin strings via ApplyEventOrigin and RemoveEventOrigin.

Discussion
---
Now that we're looking at a concrete implementation of this I already see room for improvement.

I don't like SourceType in practice, for a couple of reasons:
 - InputManager.DetermineInputSourceType seems like it could be _extremely_ flaky even if we add Handedness to IInputSource.  
 - I initially pushed to keep SourceType out of IInputSource, but it looks like these additional MixedRealityInputType types could be incorporated into IInputSource without much disruption.
 - Determining the input source type via IInputSource is also redundant when we're getting events directly from the device type in question (eg RaiseTouchpadTouched / RaiseGamePadDetected).

So my feeling is we should move this information into IInputSource after all.

Using the original Tag object
---
I played with using Tag instead of the EventOrigin string. On the InputManager side this wasn't difficult. Any existing uses would definitely be broken, but that may not be a big deal.

Consuming these tags is also less trouble than expected. BUT there's an interdependence problem I didn't notice until I implemented a few button types:

```
// unsafe examples

public class ButtonString : MonoBehaviour, IInputClickHandler {
        public void OnInputDown(InputEventData eventData)
        {
                string tagAsString = (string)eventData.Tag;
                if (tagAsString == someDefinedString) {
                        // do thing
                }
        }
}

public class ButtonGameObject : MonoBehaviour, IInputClickHandler {
        public void OnInputDown(InputEventData eventData)
        {
                GameObject tagAsGameObject = (GameObject )eventData.Tag;
                tagAsGameObject.GetComponent<SomeKindOfPointer>().DoThing(); 
        }
}

public class ButtonStruct : MonoBehaviour, IInputClickHandler {
        public void OnInputDown(InputEventData eventData)
        {
                MyStruct tagAsMyStruct = (MyStruct)eventData.Tag;     
                if (tagAsMyStruct.someBoolValue) {
                        // do thing
                }           
        }
}

```

On the button end you'll have to use exception handling when casting tags, and you'll have to implement fail-safe behaviors for tags that aren't the right type & can't be consumed. Maybe not a big deal.

On the pointer end, sending consumable tags to all three button types would require knowledge of the kind of button it's looking at. So you've got a choice between very complex pointers, or pointers that only work with a specific kind of button. This interdependence is a bigger deal. It sort of undermines the generic nature of the input system.

We could suggest that devs stick to a single tag type in their project to keep things simple, but then there's little point in using a generic tag object.

So it looks like a fixed type is the way to go. A string still seems like the obvious choice, but if anyone has thoughts on using an int / GameObject / whatever, please chime in.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/InputEventData/BaseInputEventData.cs:
```diff
@@ -5,6 +5,21 @@
 
 namespace HoloToolkit.Unity.InputModule
 {
+    public enum MixedRealityInputType
```
I think we could use a better name.  Mixed Reality is generally for VR/AR experiences, but we have touchscreen here?

Railboy replied:
Agreed, and this enum may not even be necessary (see note on incorporating into IInputHandler).

There's a big push for Touchscreen in MR. Asymmetrical multiplayer, observers, etc. Haven't seen many examples but it's considered part of the ecosystem.

StephenHodgson replied:
Yeah I agree, we should add to the enum that's already in the `IInputHandler`

Railboy replied:
I've added these to the InteractionSourceInfo.

This breaks that enum's parity with Unity's UnityEngine.XR.WSA.Input.InteractionSourceKind but there are no obvious consequences.

StephenHodgson replied:
I think it should be fine.  I don't see them changing that enum anytime soon anyway. (But then again it is Unity)

StephenHodgson commented on UnityPackageManager/manifest.json:
```diff
@@ -0,0 +1,4 @@
+{
+	"dependencies": {
```
please remove.  I believe this was added to the `.gitignore` in a different PR

Railboy replied:
Good, this file's been driving me nuts.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Managers/InputManager.cs:
```diff
@@ -61,6 +61,146 @@ public class InputManager : Singleton<InputManager>
+            if (!inputSourceTypes.ContainsKey(newSource.SourceId)) {
+                inputSourceTypes.Add(newSource.SourceId, inputType);
+            } else {
+                inputSourceTypes[newSource.SourceId] = inputType;
```
I don't think there's currently anything that checks to make sure new detected sources have the same Id, and potentially different sources can have the same id even though the source kind is different.

Railboy replied:
@keveleigh warned me about about this. Do we have a cheat sheet for expected ID behavior?

StephenHodgson replied:
None, that I'm aware of.

I think we should have the InputManager assign this sourceId as they're detected.

Railboy replied:
I've removed these functions along with SourceType.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/InputEventData/BaseInputEventData.cs:
```diff
@@ -25,16 +40,28 @@ public abstract class BaseInputEventData : BaseEventData, IInputSourceInfoProvid
+        /// <summary>
+        /// An optional, input-source-dependent tag to be associated with this input source
+        /// </summary>
+        public MixedRealityInputType SourceType { get; private set; }
```
Isn't the source kind part of the `IInputSource`?

Should we just get it from there?

Railboy replied:
Probably. See my notes above.

>my feeling is we should move this information into IInputSource after all.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Managers/InputManager.cs:
```diff
@@ -233,15 +373,15 @@ private void Start()
         #endregion // Unity APIs
 
-        public void HandleEvent<T>(BaseEventData eventData, ExecuteEvents.EventFunction<T> eventHandler) where T : IEventSystemHandler
+        public void HandleEvent<T>(BaseInputEventData eventData, ExecuteEvents.EventFunction<T> eventHandler) where T : IEventSystemHandler
```
What's the need/benefit to change this from the more generic `BaseEventData`?

Railboy replied:
None (yet) - I'll revert this for now.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Managers/InputManager.cs:
```diff
@@ -61,6 +61,146 @@ public class InputManager : Singleton<InputManager>
+        private string defaultEventOrigin = string.Empty;
+
+        private void ApplyDefaultEventOrigin(InputSourceInfo newSource) {
+            return;
```
unfinished?

Railboy replied:
Unnecessary, part of an earlier experiment.

Railboy commented:
Changes
---

Removed SourceType from BaseInputEventData.

Moved MixedRealityInputType values to InteractionSourceInfo.

Created non-WSA InteractionSourceHandedness enum for use in IInputSource interface.

Added TryGetSourceHandedness to IInputSource interface + all input source scripts.

Railboy commented:
Changes
---

Added MultiPointerSelector to InputManager prefab.

Added CursorOverride to IPointingSource.

FocusManager now provides a list of active IPointingSources via PointingSources.

StephenHodgson commented on Assets/HoloToolkit/UX/Scripts/Pointers/ControllerPointerBase.cs:
```diff
@@ -0,0 +1,166 @@
+
+namespace HoloToolkit.Unity.UX
+{
+    public abstract class ControllerPointerBase : MonoBehaviour, IPointingSource
```
Should this actually be in UX or InputModule?

StephenHodgson commented on Assets/HoloToolkit/UX/Scripts/Pointers/ControllerPointerBase.cs:
```diff
@@ -0,0 +1,166 @@
+using UnityEngine.EventSystems;
+using UnityEngine.XR.WSA.Input;
+
+namespace HoloToolkit.Unity.UX
```
Should this actually be in UX or InputModule?

Railboy replied:
Not sure. There will be multiple pointer types with different looks, shapes, capabilities, etc. So we're kind of veering into design / UX territory rather than core functionality.

StephenHodgson commented on Assets/HoloToolkit/UX/Scripts/Pointers/ControllerPointerBase.cs:
```diff
@@ -0,0 +1,166 @@
+using System;
+using UnityEngine;
+using UnityEngine.EventSystems;
+using UnityEngine.XR.WSA.Input;
```
No WSA platform checks

StephenHodgson commented on Assets/HoloToolkit/UX/Scripts/Pointers/FocusPointerInput.cs:
```diff
@@ -0,0 +1,119 @@
+using UnityEngine.VR.WSA.Input;
+#endif
+
+namespace HoloToolkit.Unity.UX
```
I think this also should belong in InputModule namespace

Railboy commented:
When you get a chance take a look at our implementation of IInputSource handedness:

` bool TryGetSourceHandedness(uint sourceId, out InteractionSourceHandedness sourceHandedness);`

To keep things non-WSA friendly we created a new **InteractionSourceHandedness** enum in HoloToolkit.Unity.InputModule to match the one from XR.WSA.Input.InteractionSourceHandedness, similar to InteractionSourceInfo / InteractionSourceKind.

This occasionally makes checking handedness awkward since you have to cast WSA handedness to HoloToolkit handedness when comparing info from the raw source with info from an IInputSource. But maybe that's unavoidable?

Also, should we come up with a different name to prevent confusion?

Railboy commented:
Changes
---

Added a parabolic navigation pointer as well as navigation hotspots / teleport pad.

Also added PointerTeleportManager, which currently conflicts with MixedRealityTeleport - I'd like to pull the functionality from PointerTeleportManager into MixedRealityTeleport before merging this PR. @keveleigh could use your suggestions here.

Railboy commented:
IFocusable -> IFocusTarget
----

**IFocusable** has been replaced with the multi-pointer-friendly **IFocusTarget**.

Focus events have been broken into two categories, **FocusEnter / Exit** and **FocusGained / Lost**, where the former _may or may not_ result in the overall loss or gain of focus, and the latter definitely indicates overall loss or gain of focus. FocusGained / Lost are global events raised by FocusManager, while FocusEnter / Exit events are raised by InputManager and handled by IFocusTarget.

Because the IFocusTarget interface is slightly more complex than the single-pointer IFocusable we've created a base **FocusTarget** class for convenience. It's not necessary to inherit from this class to implement IFocusTarget, but it does reduce some of the boilerplate code necessary to manage the focus target's Focusers list.

The purpose served by IPointerSpecificFocusHandler has been largely absorbed by IFocusTarget so it has been renamed to **IFocusChangedHandler** and contains only one method, **OnFocusChanged**. This method receives event data containing the IFocuser, previous and current focus target GameObjects. We should probably turn this into a global OnFocusChanged event rasied by FocusManager as well.

Railboy commented:
IFocuser, IPointingSource & PointerResult
----

`GameObject Target` and `GameObject PreviousTarget` have been moved out of PointerResult (previously PointerData) and into the **IFocuser** interface:

```
public interface IFocuser
{
    bool InteractionEnabled { get; }
    bool FocusLocked { get; set; }    
    bool OwnsInput(BaseInputEventData eventData);
    GameObject Target { get; set; }
    GameObject PreviousTarget { get; set; }
}
```

This enables us to move `PointerResult Result` (which contains data not relevant to generic focusers) out of IFocuser and into IPointingSource.

PointerResult retains a reference to its IPointingSource (and thus its Target / PreviousTarget) so its functionality is largely unchanged. However, it is now a struct to reinforce the idea that FocusManager is responsible for managing its contents. When a focuser is registered in FocusManager, pointing sources are given a newly initialized PointerData.

Placing Target and PreviousTarget in IFocuser has the added benefit of giving developers greater control over focus handling. FocusManager should be solely responsible for managing the data in PointerResult, but a focuser should have the power to override or veto FocusManager's attempts to set Target and PreviousTarget.

Setting Target and PreviousTarget internally (or preventing them from being set externally) will not interfere with focus events being raised normally.

StephenHodgson commented:
@Railboy let's address some of these conflicts

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Focus/IFocusTarget.cs:
```diff
@@ -0,0 +1,25 @@
+using HoloToolkit.Unity.InputModule;
+using System.Collections.Generic;
+
+public interface IFocusTarget : IEventSystemHandler
```
Why not keep to the usual naming conventions and make this `IFocusHandler`?

Railboy replied:
Being the _target_ of focus is different than a thing _handling_ focus. It sounds like a pointless differentiation now but once you break apart Focusers and Pointers we found it makes things clearer conceptually.

Also, the behavior of a FocusTarget is different enough from an old FocusHandler that we didn't want to invite confusion by calling them the same thing.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Focus/PointerResult.cs:
```diff
@@ -3,24 +3,122 @@
+    public struct PointerResult
     {
-        public Vector3 StartPoint { get; protected set; }
+        public PointerResult (IPointingSource pointingSource)
```
needs to initialize `this`

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Gaze/GazePointer.cs:
```diff
@@ -8,35 +8,11 @@
+    /// Gaze handles everything related to a gaze ray that can interact with other objects.
     /// </summary>
-    public class GazeManager : Singleton<GazeManager>, IPointingSource
+    public class GazePointer : Singleton<GazePointer>, IPointingSource
```
I think the rename is not needed. The usual pattern for singletons is the manager suffix.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/InputEventData/BaseInputEventData.cs:
```diff
@@ -23,18 +23,24 @@ public abstract class BaseInputEventData : BaseEventData, IInputSourceInfoProvid
         }
 
-        protected virtual void BaseInitialize(IInputSource inputSource, uint sourceId, object tag)
+        protected virtual void BaseInitialize(IInputSource inputSource, uint sourceId, string eventOrigin, object[] tag = null)
```
Still not a fan of this string.

StephenHodgson replied:
I almost feel like the source and it's id should be enough to determine what the events origin is.

Railboy replied:
It's enough to differentiate between a single device, but not between multiple sources that use a single device, which is a common situation. (You agreed that this was useful at the the meeting where it was presented so I'm going to hold you to past-Stephen's opinion.)

This is definitely something that would have to survive the transition to #1603, assuming we close this branch.

StephenHodgson replied:
I think with my refactor work, because each Input Source is unique, you should no longer get events that aren't from a single source. 

Please review and let me know if that's not the case.

Railboy replied:
It doesn't look like that's the case. Here's another way to tell: If I have 3 pointer components attached to my left controller, and I receive a click event from the InputSource associated with that controller, can your event tell me which of those 3 pointers was responsible for raising that click event?

StephenHodgson replied:
Yes, that is correct

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/InputSources/InteractionSourceInfo.cs:
```diff
@@ -5,9 +5,16 @@ namespace HoloToolkit.Unity.InputModule
+        Mouse,
+        Gamepad,
+        Keyboard,
+        TouchScreen,
```
If we're defining the enum values we should probably use the flags attribute and make the assignments bit operators

StephenHodgson replied:
Also, it seems like some of these changes are beyond the scope of just multi pointers

Railboy replied:
Yeah we can drop this stuff.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Managers/MixedRealityTeleport.cs:
```diff
@@ -119,10 +119,11 @@ private void HandleGamepad()
                 if (currentPointingSource == null && leftY > 0.8 && Math.Abs(leftX) < 0.3)
                 {
-                    if (FocusManager.Instance.TryGetSinglePointer(out currentPointingSource))
+                    throw new NotImplementedException();
```
Def want to make sure this is fixed before merging.

StephenHodgson commented on Assets/HoloToolkit-Examples/SpatialUnderstanding/Scripts/SpatialUnderstandingBasicCursor.cs:
```diff
@@ -48,14 +48,14 @@ protected virtual void Awake()
         protected virtual void Start()
         {
-            gazeManager = GazeManager.Instance;
+            gaze = GazePointer.Instance;
```
Not sure if it's worth caching the reference if we're just going to call it directly below

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs:
```diff
@@ -44,16 +44,18 @@ protected override void Awake()
                     }
 
-                    RegisterPointer(pointingSource);
+                    RegisterFocuser(pointingSource);
```
I think don't think there's a strong reason to rename Pointers to Foucusers.

Railboy replied:
We're not renaming them, we're splitting them into two distinct entities. Focusers can focus on a thing without a ray. Eg, by touching them, or whatever. Pointers are a kind of Focuser.

This change has to come sooner or later.

StephenHodgson replied:
Ah I see. Well why not just use the built in `Input.Selectable` to handle this kinda stuff?

StephenHodgson replied:
I don't see another place where we're registering focusers that arent' pointers, or is this in a later revision?

Railboy replied:
It's in the latest revision. RegisterFocuser (line 172) now takes an actual IFocuser instead of a pointingSource. I haven't changed the registeredPointers array to a registeredFocusers array yet, so that naming might cause confusion.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs:
```diff
@@ -44,16 +44,18 @@ protected override void Awake()
                 }
             }
+
+            pointerInputEventData = new PointerInputEventData(EventSystem.current);
```
Event datas should only ever be created in the InputManager and handed there.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs:
```diff
@@ -282,165 +211,77 @@ public void RegisterPointer(IPointingSource pointingSource)
+        /// Called when a focus target with focus loses focus from ALL focusers
+        /// </summary>
+        public delegate void FocusLostMethod(IFocusTarget unfocusedObject);
+        public event FocusLostMethod FocusLost;
```
There's no reason to have these events if we're already raising them through the InputManager

Railboy replied:
These were moved from InputManager to FocusManager for clarity.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Focus/InputSourcePointer.cs:
```diff
@@ -19,6 +19,8 @@ public class InputSourcePointer : IPointingSource
 
         public BaseRayStabilizer RayStabilizer { get; set; }
 
+        public Cursor CursorOverride { get; set; }
```
I don't think this should be an override, but instead have the InputSourcePointer own the Cursor

Railboy replied:
With MultiPointerSelector.cs (as opposed to SimpleSinglePointerSelector.cs) it's a little more intuitive to opt-in to unique cursors. But either way works.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Focus/InputSourcePointer.cs:
```diff
@@ -34,6 +36,10 @@ public RayStep[] Rays
 
         public PointerResult Result { get; set; }
 
+        public GameObject Target { get; set; }
```
Not convinced we need this here.  It should be enough that we pass this info through the raise event

Railboy replied:
Read the section **IFocuser, IPointingSource & PointerResult** where I go over this change in detail.

StephenHodgson commented:
@Railboy is there any way you can update this PR with the latest changes from the working dev branch?

Railboy commented:
You'll have to overhaul these changes as you incorporate them into your branch anyway, so there's no point in changing everything twice.

StephenHodgson commented:
@Railboy should we go ahead and close this?

DDReaper commented:
Changes merged in to the InputSystem refactor
